clients/drcachesim/tracer/tracer.cpp

```diff
@@ -669,12 +752,61 @@ event_pre_syscall(void *drcontext, int sysnum);
 static void
 event_kernel_xfer(void *drcontext, const dr_kernel_xfer_info_t *info);
 
+// Returns whether we've reached the end of this tracing window.
+static bool
+count_traced_instrs(void *drcontext, int toadd)
+{
+    per_thread_t *data = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    data->instr_count += toadd;
+    if (data->instr_count >= local_instr_count_threshold()) {
+        uint64 newval =
+            traced_instr_count.fetch_add(data->instr_count, std::memory_order_release) +
+            // fetch_add returns old value.
+            data->instr_count;
+        data->instr_count = 0;
+        if (newval >= op_trace_for_instrs.get_value())
+            return true;
+    }
+    return false;
+}
+
+static void
+reached_traced_instrs_threshold(void *drcontext)
+{
+    per_thread_t *data = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    dr_mutex_lock(mutex);
+    if (get_local_window(data) != tracing_window.load(std::memory_order_acquire)) {
+        // Another thread already changed the mode.
+        dr_mutex_unlock(mutex);
+        return;
+    }
+    // We've reached the end of our window.
+    // We do not attempt a proactive synchronous flush of other threads'
+    // buffers, relying on our end-of-block check for a mode change.
+    // (If -trace_every_instrs is not set and we're not going to trace
+    // again, we still use a counting mode for simplicity of not adding
```

Comment:
I was considering re-implementing -max_global_trace_refs as -trace_for_instrs without -retrace_every_instrs, since the current max keeps tracing but just doesn't output which doesn't seem ideal; plus it counts refs while instrs seems better.  Maybe we can deprecate it.  I'm going to add that to the list of cleanup/optional fixes for after this PR: this PR is big enough as it is.  Maybe I will file a separate issue, or at least list them in #3995.